src/dstack/_internal/gateway/services/service_proxy.py

```diff
@@ -0,0 +1,157 @@
+from typing import AsyncGenerator, AsyncIterator, Optional
+
+import fastapi
+import httpx
+from starlette.requests import ClientDisconnect
+
+from dstack._internal.gateway.repos.base import BaseGatewayRepo
+from dstack._internal.gateway.services.service_connection import service_replica_connection_pool
+from dstack._internal.utils.logging import get_logger
+
+logger = get_logger(__name__)
+
+
+async def proxy(
+    project_name: str,
+    run_name: str,
+    path: str,
+    request: fastapi.Request,
+    repo: BaseGatewayRepo,
+) -> fastapi.responses.Response:
+    if "Upgrade" in request.headers:
+        raise fastapi.exceptions.HTTPException(
+            fastapi.status.HTTP_400_BAD_REQUEST, "Upgrading connections is not supported"
+        )
+
+    service = await repo.get_service(project_name, run_name)
+    if service is None or not service.replicas:
+        raise fastapi.HTTPException(
+            fastapi.status.HTTP_404_NOT_FOUND,
+            f"Service {project_name}/{run_name} not found",
+        )
+    if service.auth:
+        # TODO(#1595): support auth
+        raise fastapi.HTTPException(
+            fastapi.status.HTTP_400_BAD_REQUEST,
+            f"Service {project_name}/{run_name} requires auth, which is not yet supported",
+        )
+
+    replica = random.choice(service.replicas)
+
+    connection = await service_replica_connection_pool.get(replica.id)
+    if connection is None:
+        project = await repo.get_project(project_name)
+        if project is None:
+            raise RuntimeError(f"Expected to find project {project_name} but could not")
+        connection = await service_replica_connection_pool.add(project, service, replica)
+
+    client = await connection.client()
+
+    try:
+        upstream_request = await build_upstream_request(request, path, client, replica.id)
+    except ClientDisconnect:
+        logger.debug(
+            "Downstream client disconnected before response was sent for %s %s",
+            request.method,
+            request.url,
+        )
+        raise fastapi.HTTPException(fastapi.status.HTTP_400_BAD_REQUEST, "Client disconnected")
+
+    try:
+        upstream_response = await client.send(upstream_request, stream=True)
+    except httpx.RequestError as e:
+        logger.debug(
+            "Error requesting %s %s for replica %s: %r",
+            upstream_request.method,
+            upstream_request.url,
+            replica.id,
+            e,
+        )
+        if isinstance(e, TimeoutError):
+            raise fastapi.HTTPException(fastapi.status.HTTP_504_GATEWAY_TIMEOUT)
+        raise fastapi.HTTPException(fastapi.status.HTTP_502_BAD_GATEWAY)
+
+    return fastapi.responses.StreamingResponse(
+        stream_response(upstream_response, replica.id),
+        status_code=upstream_response.status_code,
+        headers=upstream_response.headers,
+    )
+
+
+async def stream_response(
+    response: httpx.Response, replica_id: str
+) -> AsyncGenerator[bytes, None]:
```

Comment:
I think `AsyncIterator[YieldType]` could be used in place of `AsyncGenerator[YieldType, None]`

Reply:
I subjectively prefer more generic return types in public interfaces and more specific return types in private ones. `AsyncGenerator` is more specific than `AsyncIterator` here